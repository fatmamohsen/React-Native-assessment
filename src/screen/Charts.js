import React, {useState} from 'react';
import {View, Text, StyleSheet, TouchableOpacity} from 'react-native';
import RNPickerSelect from 'react-native-picker-select';
import {LineChart, Grid} from 'react-native-svg-charts';
import axios from 'axios';
const API_KEY = '0bbdd6fc55-a9ae53a12a-rxjd4t';
const mokeData = [50, 90, 30, 100, 70, 10];

export const Charts = () => {
  const [chartData, setChartData] = useState(null);
  const [fromCurrency, setFromCurrency] = useState('USD');
  const [toCurrency, setToCurrency] = useState('EUR');
  const [timeInterval, setTimeInterval] = useState('1D');

  const fetchChartData = async () => {
    try {
      const response = await axios.get(
        `https://api.exchangeratesapi.io/history?start_at=${getStartDate()}&end_at=${getEndDate()}&base=${fromCurrency}&symbols=${toCurrency}&access_key=${API_KEY}`,
      );
      const data = response.data;
      console.log(data);
      if (data && data.rates) {
        const formattedData = formatChartData(data);
        setChartData(formattedData);
      } else {
        setChartData(mokeData);
        console.error('Error retrieving exchange rate data');
      }
    } catch (error) {
      console.error(error);
    }
  };

  const getStartDate = () => {
    const now = new Date();
    const startDate = new Date();
    switch (timeInterval) {
      case '1D':
        startDate.setDate(now.getDate() - 1);
        break;
      case '1M':
        startDate.setMonth(now.getMonth() - 1);
        break;
      case '3M':
        startDate.setMonth(now.getMonth() - 3);
        break;
      case '1Y':
        startDate.setFullYear(now.getFullYear() - 1);
        break;
      case '5Y':
        startDate.setFullYear(now.getFullYear() - 5);
        break;
    }
    return formatDate(startDate);
  };

  const getEndDate = () => formatDate(new Date());

  const formatDate = date => date.toISOString().slice(0, 10);

  const formatChartData = data => {
    const labels = Object.keys(data.rates);
    const rates = labels.map(label => data.rates[label][toCurrency]);
    return rates;
  };

  return (
    <View style={styles.container}>
      <Text style={styles.label}>From Currency:</Text>
      <RNPickerSelect
        value={fromCurrency}
        onValueChange={value => setFromCurrency(value)}
        items={[
          {label: 'US Dollar', value: 'USD'},
          {label: 'Euro', value: 'EUR'},
          {label: 'British Pound', value: 'GBP'},
          {label: 'Japanese Yen', value: 'JPY'},
          {label: 'Canadian Dollar', value: 'CAD'},
        ]}
        placeholder={{
          label: 'Select a currency...',
          value: null,
        }}
        style={{
          inputIOS: styles.picker,
          inputAndroid: styles.picker,
        }}
      />
      <Text style={styles.label}>To Currency:</Text>
      <RNPickerSelect
        value={toCurrency}
        onValueChange={value => setToCurrency(value)}
        items={[
          {label: 'Euro', value: 'EUR'},
          {label: 'US Dollar', value: 'USD'},
          {label: 'British Pound', value: 'GBP'},
          {label: 'Japanese Yen', value: 'JPY'},
          {label: 'Canadian Dollar', value: 'CAD'},
        ]}
        placeholder={{
          label: 'Select a currency...',
          value: null,
        }}
        style={{
          inputIOS: styles.picker,
          inputAndroid: styles.picker,
        }}
      />
      <Text style={styles.label}>Time Interval:</Text>
      <RNPickerSelect
        value={timeInterval}
        onValueChange={value => setTimeInterval(value)}
        items={[
          {label: '1 Day', value: '1D'},
          {label: '1 Month', value: '1M'},
          {label: '3 Months', value: '3M'},
          {label: '1 Year', value: '1Y'},
          {label: '5 Years', value: '5Y'},
        ]}
        placeholder={{
          label: 'Select a time interval...',
          value: null,
        }}
        style={{
          inputIOS: styles.picker,
          inputAndroid: styles.picker,
        }}
      />
      <TouchableOpacity style={styles.button} onPress={fetchChartData}>
        <Text style={styles.buttonText}>FetchData</Text>
      </TouchableOpacity>
      {chartData ? (
        <LineChart
          style={{height: 200, width: 300}}
          data={chartData}
          svg={{stroke: 'rgb(134, 65, 244)'}}
          contentInset={{top: 20, bottom: 20}}>
          <Grid />
        </LineChart>
      ) : (
        <Text style={styles.loadingText}>Loading chart data...</Text>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'white',
  },
  label: {
    fontSize: 18,
    fontWeight: 'bold',
    color: '#fff',
    marginTop: 20,
    marginBottom: 10,
  },
  picker: {
    height: 50,
    width: 200,
    backgroundColor: '#fff',
    borderRadius: 10,
    marginBottom: 20,
    paddingHorizontal: 10,
    alignSelf: 'center',
    borderWidth: 1,
  },
  button: {
    backgroundColor: '#4CAF50',
    borderRadius: 10,
    padding: 10,
    marginTop: 20,
  },
  buttonText: {
    color: '#fff',
    fontSize: 18,
    fontWeight: 'bold',
    textAlign: 'center',
  },
  loadingText: {
    color: '#fff',
    fontSize: 18,
    fontWeight: 'bold',
    textAlign: 'center',
    marginTop: 20,
  },
});
