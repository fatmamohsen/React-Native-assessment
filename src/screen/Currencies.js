import axios from 'axios';
import React, {useEffect, useState} from 'react';
import {RefreshControl, StyleSheet, Text, View, FlatList} from 'react-native';

const Item = ({item}) => {
  return (
    <View style={styles.item}>
      <Text style={styles.itemText}>{item.currency}</Text>
      <Text style={styles.itemText}>{item.rate}</Text>
    </View>
  );
};
export const CurrenciesScreen = () => {
  const [currencies, setCurrencies] = useState([]);
  const fetchCurrencies = () => {
    return axios
      .get('https://api.vatcomply.com/rates')
      .then(response => {
        const rates = response.data.rates;
        const ratesArray = Object.entries(rates).map(([currency, rate]) => ({
          currency,
          rate,
        }));
        setCurrencies(ratesArray);
      })
      .catch(error => {
        console.log(error);
      });
  };

  useEffect(() => {
    fetchCurrencies();
    const timeInterval = 10000;
    let intervalId = null;
    intervalId = setInterval(fetchCurrencies, timeInterval);
    return () => clearInterval(intervalId);
  }, []);

  return (
    <View style={styles.container}>
      <FlatList
        data={currencies}
        renderItem={({item}) => <Item item={item} />}
        showsVerticalScrollIndicator={false}
        ListEmptyComponent={() => {
          return <Text>no currencies</Text>;
        }}
        refreshControl={
          <RefreshControl refreshing={false} onRefresh={fetchCurrencies} />
        }
      />
    </View>
  );
};
const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'white',
  },
  item: {
    flexDirection: 'row',
    marginHorizontal: 24,
    justifyContent: 'space-between',
    height: 50,
    marginBottom: 16,
    alignItems: 'center',
  },
  itemText: {
    fontSize: 18,
    fontWeight: '600',
  },
});
