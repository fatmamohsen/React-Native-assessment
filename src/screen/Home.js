import React from 'react';
import {StyleSheet, Text, View} from 'react-native';

export const Home = () => {
  return (
    <View style={styles.container}>
      <Text style={styles.text}>home</Text>
    </View>
  );
};
const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'red',
    justifyContent: 'center',
  },
  text: {
    fontSize: 50,
    color: 'black',
  },
});
