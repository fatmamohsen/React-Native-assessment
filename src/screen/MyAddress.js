import React, {useState, useEffect} from 'react';
import {View, StyleSheet, Text} from 'react-native';
import Geolocation from '@react-native-community/geolocation';
import RNPickerSelect from 'react-native-picker-select';

export const MyAddressesScreen = () => {
  const [addressesList, setAddressesList] = useState([]);
  const [isAllowedLocation, setIsAllowedLocation] = useState(false);
  const [selectedAddresses, setSelectedAddresses] = useState(null);
  useEffect(() => {
    Geolocation.getCurrentPosition(
      position => {
        const {latitude, longitude} = position.coords;
        fetch(
          `https://maps.googleapis.com/maps/api/geocode/json?latlng=${latitude},${longitude}&key=YOUR_API_KEY`,
        )
          .then(response => {
            response.json();
          })
          .then(data => {
            const formattedCountries = data.map(country => ({
              label: country.name.common,
              value: country.cca2,
            }));
            setAddressesList(formattedCountries);
          })
          .catch(error => console.log(error));
      },
      error => {
        console.log(error);
      },
    );
  }, []);

  useEffect(() => {
    fetch('https://restcountries.com/v3.1/all')
      .then(response => response.json())
      .then(data => {
        const formattedCountries = data.map(country => ({
          label: country.name.common,
          value: country.cca2,
        }));
        setAddressesList(formattedCountries);
      });
  }, []);

  return (
    <View style={styles.container}>
      <Text style={styles.label}>From Currency:</Text>
      <RNPickerSelect
        value={selectedAddresses}
        onValueChange={value => {
          setSelectedAddresses(value);
        }}
        items={addressesList}
        placeholder={{
          label: 'Where are you from...',
        }}
        style={{
          inputIOS: styles.picker,
          inputAndroid: styles.picker,
        }}
      />
      {selectedAddresses ? <Text>{selectedAddresses}</Text> : null}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 24,
    backgroundColor: 'white',
  },
  label: {
    fontSize: 18,
    fontWeight: 'bold',
    color: 'black',
    marginTop: 20,
    marginBottom: 10,
    paddingHorizontal: 30,
  },
  picker: {
    height: 50,
    width: 300,
    backgroundColor: '#fff',
    borderRadius: 10,
    marginBottom: 20,
    paddingHorizontal: 10,
    alignSelf: 'center',
    borderWidth: 1,
  },
});
