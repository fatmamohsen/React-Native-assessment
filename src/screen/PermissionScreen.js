import Geolocation from '@react-native-community/geolocation';
import React from 'react';
import {Linking, Pressable, StyleSheet, Text, View} from 'react-native';
import Toast from 'react-native-toast-message';

export const MyPermission = () => {
  const showToast = ({text, type}) => {
    Toast.show({
      type: text,
      text2: type,
    });
  };
  const onRequestLocationPermissionClick = () => {
    Geolocation.requestAuthorization(
      () => {
        console.log('Geolocation permission');
      },
      error => {
        if (error.PERMISSION_DENIED === 1) {
          showToast(
            'error',
            'Please activate your location to accurately locate your position',
          );
        }
      },
    );
  };
  const allowNotifications = () => {
    Linking.openSettings();
  };
  return (
    <View style={styles.container}>
      <Text style={styles.text}>Please allow your location</Text>
      <Pressable style={styles.btn} onPress={onRequestLocationPermissionClick}>
        <Text style={styles.btnText}>Allow location</Text>
      </Pressable>
      <Text style={styles.text}>Please enable sending notificatons to you</Text>
      <Pressable style={styles.btn} onPress={allowNotifications}>
        <Text style={styles.btnText}>Allow notifications</Text>
      </Pressable>
    </View>
  );
};
const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'white',
    justifyContent: 'center',
  },
  text: {
    fontSize: 24,
    textAlign: 'center',
    color: 'black',
    marginBottom: 24,
  },
  btn: {
    marginHorizontal: 24,
    height: 48,
    borderRadius: 16,
    backgroundColor: 'black',
    marginBottom: 30,
  },
  btnText: {
    fontSize: 20,
    textAlign: 'center',
    color: 'white',
    marginVertical: 10,
  },
});
