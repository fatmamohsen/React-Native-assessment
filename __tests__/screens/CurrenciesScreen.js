import React from 'react';
import {render} from '@testing-library/react-native';
import {NavigationContainer} from '@react-navigation/native';
import {CurrenciesScreen} from '../../src/screen/Currencies';

describe('test CurrenciesScreen', () => {
  afterAll(() => jest.restoreAllMocks());

  it('should match snapshot of rendered CurrenciesScreen', () => {
    const {toJSON} = render(
      <NavigationContainer>
        <CurrenciesScreen />
      </NavigationContainer>,
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
