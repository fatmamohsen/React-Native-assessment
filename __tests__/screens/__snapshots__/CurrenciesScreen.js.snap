// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test CurrenciesScreen should match snapshot of rendered CurrenciesScreen 1`] = `
<View
  style={
    {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <RCTRefreshControl />
    <View>
      <Text>
        no currencies
      </Text>
    </View>
  </RCTScrollView>
</View>
`;
